python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -156,1253 +120,308 @@ class MakeBrighterFatterKernelTaskConfig(pexConfig.Config):
+        bfk.gain = inputPtc.gain
+        bfk.noise = inputPtc.noise
+        bfk.meanXcorrs = dict()
+        bfk.valid = dict()
 
-    def _applyGains(self, means, xcorrs, ptcData):
-        """Apply the gains calculated by the PtcTask.
+        for amp in detector:
+            ampName = amp.getName()
+            mask = np.array(inputPtc.expIdMask[ampName], dtype=bool)
+
+            gain = bfk.gain[ampName]
+            fluxes = np.array(bfk.means[ampName])[mask]
+            variances = np.array(bfk.variances[ampName])[mask]
+            xCorrList = [np.array(xcorr) for xcorr in bfk.rawXcorrs[ampName]]
+            xCorrList = np.array(xCorrList)[mask]
+
+            if gain <= 0:
+                # We've received very bad data.
+                self.log.warn("Impossible gain recieved from PTC for %s: %f.  Skipping amplifier.",
+                              ampName, gain)
+                bfk.meanXcorrs[ampName] = np.zeros(bfk.shape)
+                bfk.ampKernels[ampName] = np.zeros(bfk.shape)
+                bfk.valid[ampName] = False
+                continue
 
-        It also removes datapoints that were thrown out in the PTC algorithm.
+            fluxes = np.array([flux*gain for flux in fluxes])  # Now in e^-
+            variances = np.array([variance*gain*gain for variance in variances])  # Now in e^2-
+
+            # This should duplicate the else block in generateKernel@L1358,
+            # which in turn is based on Coulton et al Equation 22.
+            scaledCorrList = list()
+            for xcorrNum, (xcorr, flux, var) in enumerate(zip(xCorrList, fluxes, variances), 1):
+                q = np.array(xcorr) * gain * gain  # xcorr now in e^-
+                q *= 2.0  # Remove factor of 1/2 applied in PTC.
+                self.log.info("Amp: %s %d/%d Flux: %f  Var: %f  Q(0,0): %g  Q(1,0): %g  Q(0,1): %g",
+                              ampName, xcorrNum, len(xCorrList), flux, var, q[0][0], q[1][0], q[0][1])
+
+                # Normalize by the flux, which removes the (0,0)
+                # component attributable to Poisson noise.
+                q[0][0] -= 2.0*(flux)
+
+                if q[0][0] > 0.0:
+                    self.log.warn("Amp: %s %d skipped due to value of (variance-mean)=%f",
+                                  ampName, xcorrNum, q[0][0])
+                    continue
 
-        Parameters
-        ----------
-        means : `dict` [`str`, `list` of `tuple`]
-            Dictionary, keyed by ampName, containing a list of the means for
-        each visit pair.
+                q /= -2.0*(flux**2)
+                scaled = self._tileArray(q)
 
-        xcorrs : `dict` [`str`, `list` of `np.array`]
-            Dictionary, keyed by ampName, containing a list of the
-        cross-correlations for each visit pair.
+                xcorrCheck = np.abs(np.sum(scaled))/np.sum(np.abs(scaled))
+                if (xcorrCheck > self.config.xcorrCheckRejectLevel) or not (np.isfinite(xcorrCheck)):
```

Comment:
Why is `2` the default value for `self.config.xcorrCheckRejectLevel`? If this is strictly checking the triangle inequality, as it says in the comment, shouldn't it be `1` (`|A + B| <= |A| + |B|`)? Or what am I missing? Wy is it even an option?

Reply:
I added the triangle inequality message when trying to make sense of what this check is doing (previously it was called the "sum of the xcorr", which isn't right either).  Digging through the commit history, it appears this is designed to only catch impossibly bad data.  I'm not sure that it's actually possible to fail this test, which suggests it can probably be dropped.